From 53fd7f1b5caae1721c579653e56b4509e0d59568 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Tue, 24 Mar 2020 19:08:55 +0100
Subject: [PATCH 0626/1010] uapi/evl: sched: make interface C++ friendly

EVL uapi headers are pulled unmodified into user code, make sure we
can use them in C++ applications.

NOTE: those changes do NOT affect the ABI.

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 include/uapi/evl/sched.h | 60 ++++++++++++++++++++++------------------
 kernel/evl/control.c     | 10 +++----
 kernel/evl/sched/tp.c    |  2 +-
 3 files changed, 39 insertions(+), 33 deletions(-)

diff --git a/include/uapi/evl/sched.h b/include/uapi/evl/sched.h
index 4710e71edb..80147c0a82 100644
--- a/include/uapi/evl/sched.h
+++ b/include/uapi/evl/sched.h
@@ -18,25 +18,27 @@
 
 #define sched_rr_quantum	sched_u.rr.__sched_rr_quantum
 
-struct __sched_rr_param {
+struct __evl_rr_param {
 	struct __evl_timespec __sched_rr_quantum;
 };
 
 #define SCHED_QUOTA		44
 #define sched_quota_group	sched_u.quota.__sched_group
 
-struct __sched_quota_param {
+struct __evl_quota_param {
 	int __sched_group;
 };
 
+enum evl_quota_ctlop {
+	evl_quota_add,
+	evl_quota_remove,
+	evl_quota_force_remove,
+	evl_quota_set,
+	evl_quota_get,
+};
+
 struct evl_quota_ctlparam {
-	enum {
-		evl_quota_add,
-		evl_quota_remove,
-		evl_quota_force_remove,
-		evl_quota_set,
-		evl_quota_get,
-	} op;
+	enum evl_quota_ctlop op;
 	union {
 		struct {
 			int tgid;
@@ -62,31 +64,35 @@ struct evl_quota_ctlinfo {
 #define SCHED_TP		45
 #define sched_tp_partition	sched_u.tp.__sched_partition
 
-struct __sched_tp_param {
+struct __evl_tp_param {
 	int __sched_partition;
 };
 
 #define EVL_TP_IDLE	-1	/* Idle pseudo-partition */
 
+struct __evl_tp_window {
+	struct __evl_timespec offset;
+	struct __evl_timespec duration;
+	int ptid;
+};
+
+enum evl_tp_ctlop {
+	evl_install_tp,
+	evl_uninstall_tp,
+	evl_start_tp,
+	evl_stop_tp,
+	evl_get_tp,
+};
+
 struct evl_tp_ctlparam {
-	enum {
-		evl_install_tp,
-		evl_uninstall_tp,
-		evl_start_tp,
-		evl_stop_tp,
-		evl_get_tp,
-	} op;
+	enum evl_tp_ctlop op;
 	int nr_windows;
-	struct __sched_tp_window {
-		struct __evl_timespec offset;
-		struct __evl_timespec duration;
-		int ptid;
-	} windows[0];
+	struct __evl_tp_window windows[0];
 };
 
 struct evl_tp_ctlinfo {
 	int nr_windows;
-	struct __sched_tp_window windows[0];
+	struct __evl_tp_window windows[0];
 };
 
 #define evl_tp_paramlen(__p)	\
@@ -96,9 +102,9 @@ struct evl_sched_attrs {
 	int sched_policy;
 	int sched_priority;
 	union {
-		struct __sched_rr_param rr;
-		struct __sched_quota_param quota;
-		struct __sched_tp_param tp;
+		struct __evl_rr_param rr;
+		struct __evl_quota_param quota;
+		struct __evl_tp_param tp;
 	} sched_u;
 };
 
@@ -115,7 +121,7 @@ union evl_sched_ctlinfo {
 struct evl_sched_ctlreq {
 	int policy;
 	int cpu;
-	union evl_sched_ctlparam *param;
+	const union evl_sched_ctlparam *param;
 	union evl_sched_ctlinfo *info;
 };
 
diff --git a/kernel/evl/control.c b/kernel/evl/control.c
index 10ba184be3..82662b6a3d 100644
--- a/kernel/evl/control.c
+++ b/kernel/evl/control.c
@@ -100,7 +100,7 @@ static int stop_services(void)
 
 #ifdef CONFIG_EVL_SCHED_QUOTA
 
-static int do_quota_control(struct evl_sched_ctlreq *ctl)
+static int do_quota_control(const struct evl_sched_ctlreq *ctl)
 {
 	union evl_sched_ctlparam param, __user *u_ctlp;
 	union evl_sched_ctlinfo info, __user *u_infp;
@@ -127,7 +127,7 @@ static int do_quota_control(struct evl_sched_ctlreq *ctl)
 
 #else
 
-static int do_quota_control(struct evl_sched_ctlreq *ctl)
+static int do_quota_control(const struct evl_sched_ctlreq *ctl)
 {
 	return -EINVAL;
 }
@@ -136,7 +136,7 @@ static int do_quota_control(struct evl_sched_ctlreq *ctl)
 
 #ifdef CONFIG_EVL_SCHED_TP
 
-static int do_tp_control(struct evl_sched_ctlreq *ctl)
+static int do_tp_control(const struct evl_sched_ctlreq *ctl)
 {
 	union evl_sched_ctlinfo *info = NULL, __user *u_infp;
 	union evl_sched_ctlparam param, __user *u_ctlp;
@@ -177,14 +177,14 @@ static int do_tp_control(struct evl_sched_ctlreq *ctl)
 
 #else
 
-static int do_tp_control(struct evl_sched_ctlreq *ctl)
+static int do_tp_control(const struct evl_sched_ctlreq *ctl)
 {
 	return -EINVAL;
 }
 
 #endif
 
-static int do_sched_control(struct evl_sched_ctlreq *ctl)
+static int do_sched_control(const struct evl_sched_ctlreq *ctl)
 {
 	int ret;
 
diff --git a/kernel/evl/sched/tp.c b/kernel/evl/sched/tp.c
index ef0f330efd..2d5a0444d8 100644
--- a/kernel/evl/sched/tp.c
+++ b/kernel/evl/sched/tp.c
@@ -314,7 +314,7 @@ static int tp_control(int cpu, union evl_sched_ctlparam *ctlp,
 	struct evl_tp_ctlparam *pt = &ctlp->tp;
 	ktime_t offset, duration, next_offset;
 	struct evl_tp_schedule *gps, *ogps;
-	struct __sched_tp_window *p, *pp;
+	struct __evl_tp_window *p, *pp;
 	struct evl_tp_window *w, *pw;
 	struct evl_tp_ctlinfo *it;
 	unsigned long flags;
-- 
2.30.2

